Language statistics from local repository

koffeinfrei opened the pull request:
Adds support for gitlab languages.

Fixes https://github.com/panter/dashboard.panter.ch/issues/20

Ready for review.

koffeinfrei committed:
extract base class for local repo operations

koffeinfrei committed:
languages for gitlab

koffeinfrei committed:
github languages from local repository

koffeinfrei committed:
languages = github + gitlab

koffeinfrei committed:
fix combination of two percent sets

percent + percent is no good. we need to divide by 2 to get a
proper percentage again.
also: reorder after combination.

koffeinfrei commented:
@toupeira /  @huerlisi / @exul care for a review?

huerlisi commented on lib/data_fetchers/git_fetcher.rb:
```diff
@@ -38,18 +38,31 @@ def line_changes
   def languages
-    languages = github.languages.map do |language, percent|
+    languages = github.languages
+      .merge(gitlab.languages) { |key, value1, value2| ((value1 + value2) / 2.0).round }
```
My understanding of this is that we use the average of the percentages of gitlab and github. This may be misleading as one of the hosters might have many more lines/projects than the other. If we mirror some big projects from github to gitlab it will also have quite some influence.

Not sure how to fix this, but we might:
- show both numbers seperately
- use some weights based on projects/line numbers (not sure what the statistic is based on)
- use linguist on checkouts of both github and gitlab projects instead of using gitlabs stats

The influence of having the github projects mirrored on gitlab could be quite big, but I guess we need to do some ballpark calculations to estimate this.

koffeinfrei replied:
You're right actually, summing up percentages doesn't work like this.
We could actually return the absolute numbers from github and gitlab (`Github#languages` and `GitlabClient#languages) and calculate the percentage here, then we'll have the correct average. Or am I missing something?

huerlisi replied:
Yepp, using the absolute numbers gives proper result, mathematically.

But we still need to take care not to calculate mirrored projects twice.

koffeinfrei replied:
Mirrorred repos are handled by adding them to the repo blacklist (see .env). Not a very elegant solution, but for now it's enough, as there aren't too many of them. For the cloning and updating of the local repos this needs to be handled anyway.

koffeinfrei replied:
Updated.

I'm thinking, that we should maybe change the technologies to only consider the `primary_framework` per repo. What do you think?

huerlisi commented on lib/data_fetchers/git_fetcher.rb:
```diff
@@ -38,18 +38,31 @@ def line_changes
   def frameworks
-    frameworks = github.frameworks.merge(gitlab.frameworks) { |key, value1, value2| (value1 + value2).round }
+    frameworks = github.frameworks
+      .merge(gitlab.frameworks) { |key, value1, value2| ((value1 + value2 / 2.0)).round }
```
Same issue as with languages...

koffeinfrei committed:
languages: absolute values insteaf of percent

as we'll have to combine the results from github and gitlab
we need to have the absolute values.

koffeinfrei committed:
frameworks: absolute values instead of percent

as we'll have to combine the results from github and gitlab
we need to have the absolute values.

koffeinfrei committed:
add missing require

toupeira commented:
:+1: looks good!
